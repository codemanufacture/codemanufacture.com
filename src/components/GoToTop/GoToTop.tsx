import React, { FunctionComponent, useState, useEffect } from 'react'
import styled from 'styled-components'
import Icon from './images/chevron.svg'
import classNames from 'classnames'
import { throttle } from 'throttle-debounce'
import { colors, transitions, zIndexes } from '../../theme'

const StyledGoToTop = styled.a`
  display: flex;
  opacity: 0;
  width: 40px;
  height: 40px;
  border-radius: 2px;
  position: fixed;
  z-index: ${zIndexes.goToTop};
  right: 30px;
  bottom: 30px;
  background: rgba(0, 0, 0, 0.3);
  transition:
    background ${transitions.basicTransition},
    opacity ${transitions.basicTransition};
  pointer-events: none;

  &.active {
    opacity: 1;
    pointer-events: initial;
  }

  &:hover {
    background: ${colors.brand};
  }

  img {
    margin: auto;
  }
`

const GoToTop: FunctionComponent = () => {
  const [buttonVisibility, setButtonVisibility] = useState(false)
  const handleClick = (e: React.SyntheticEvent) => {
    e.preventDefault()
    window.scroll(0, 0)
  }

  const handler = throttle(100, () => {
    if (window.scrollY > 0 && !buttonVisibility) {
      setButtonVisibility(true)
    } else if (window.scrollY === 0) {
      setButtonVisibility(false)
    }
  })

  useEffect(() => {
    window.addEventListener('scroll', handler)
    return () => window.removeEventListener('scroll', handler)
  })

  const buttonClassName = classNames({
    active: buttonVisibility,
  })

  return (
    <StyledGoToTop
      className={buttonClassName}
      title="Go to top"
      href="#go-to-top"
      onClick={e => handleClick(e)}
    >
      <img src={Icon} alt="Go to top" width={12} />
    </StyledGoToTop>
  )
}

export default GoToTop
