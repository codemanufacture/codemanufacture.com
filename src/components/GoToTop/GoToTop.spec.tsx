import * as React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import GoToTop from './GoToTop'

describe('GoToTop component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<GoToTop />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
