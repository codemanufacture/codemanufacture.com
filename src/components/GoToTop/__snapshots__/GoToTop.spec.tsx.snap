// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoToTop component renders correctly 1`] = `
<a
  className="sc-bdVaJa jkbWwW"
  href="#go-to-top"
  onClick={[Function]}
  title="Go to top"
>
  <img
    alt="Go to top"
    src={
      Object {
        "0": "t",
        "1": "e",
        "10": "s",
        "11": "t",
        "12": "u",
        "13": "b",
        "2": "s",
        "3": "t",
        "4": "-",
        "5": "f",
        "6": "i",
        "7": "l",
        "8": "e",
        "9": "-",
        "default": "test-file-stub",
      }
    }
    width={12}
  />
</a>
`;
